### DISCUSSION TITLE
How to propagate Trace id to mongodb and get the tracing details while reading from collection

### DISCUSSION THREAD

**Discussion by sulthans** [open-telemetry/opentelemetry-java-instrumentation#12496] [Q&A] [answered]
I am using Java agent version - 2.8.0, with the below environment set I am seeing traces in jaeger UI for the DB insert with collection names. 
OTEL_INSTRUMENTATION_MONGO_ENABLED=true;

This collection is reading from different scheduler application, is it possible to get the same trace context ?
I have added this env variable both applications - OTEL_PROPAGATORS=tracecontext, baggage but new trace is generated whenever app trying to read from DB.

**Comment by laurit** [comment, marked as answer]
This is the expected, trace context is not persisted in mongodb.

**Comment by trask** [comment]
@sulthans you may be interested to follow https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11940 (I'm not sure how to put a "comment" into a mongodb query, but maybe there's a way)

**Reply by sulthans** [reply]
Okay, let me explore "comment" for NoSQL DB. 
@trask @laurit I am not getting trace logs in console for Spring boot scheduler application but trace can be seen in jaeger UI.  The same is working with Spring boot application. Did i missed anything?

Environment variable used in Spring boot scheduler,

```
JAVA_TOOL_OPTIONS="-javaagent:/Users/sulthan/Documents/project/lib/opentelemetry-javaagent.jar";
OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318;
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf;
OTEL_LOGS_EXPORTER=none;
OTEL_METRICS_EXPORTER=none;
OTEL_TRACES_EXPORTER=otlp
```

Log4j2.xml : 

```
 <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
              <PatternLayout
                    pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} trace_id=%X{trace_id} span_id=%X{span_id} - 
                    %msg%n"/>
       </Console>
</Appenders>
<Loggers>
   <Root level="INFO">
            <AppenderRef ref="Console"/>
   </Root>
</Loggers>

```

**Reply by laurit** [reply]
Do you mean that you are not seeing `trace_id=` filled in your logs? If so keep in mind that this is only filled for log statements that happen in the context of an active trace.

**Reply by sulthans** [reply]
@laurit I am seeing empty trace_id= during startup. 
I have added exclude forgot to mention, unnecessary logs when no data returns in scheduler so excluded method scheduleTask. But method scheduleTask calling multiple function I have not excluded those which has logs are appearing in console without trace_id

`-Dotel.instrumentation.methods.exclude=Scheduler#scheduleTask(),Scheduler#fetchEvents(),SchedulerService#getEvents()
`

**Reply by laurit** [reply]
It is not clear what you are doing and what you are expecting. Please provide a minimal sample application that illustrates your issue along with any instructions necessary. Are you sure that `otel.instrumentation.methods.exclude` is a valid configuration option?

**Reply by sulthans** [reply]
@laurit @trask i will remove the configuration and test this. 
How to log current span id with auto instrumentation?
I have a requirement to trace with parent span where my application calls multiple service. 
If not do we have confiiguration for this or whether i can modify log4j instrumentation and add as a extension library? Could you please suggest to add current span Id in logs with minimal or without code change.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-context-data/log4j-context-data-2.17/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/log4j/contextdata/v2_17/OpenTelemetryContextDataProvider.java

**Reply by laurit** [reply]
> How to log current span id with auto instrumentation?

See the sample log4j configuration at the end of https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-context-data/log4j-context-data-2.17/library-autoconfigure/README.md This instrumentation is included in the agent you don't need to do anything besides changing the logging pattern.

**Reply by sulthans** [reply]
@laurit Thanks for reply, i may wrong with my question. Current span which i mean from where i am making multiple API calls.
This log4j2.xml config one which is printing the current span but i need to log the parent_id please refer below for which has parent span as parent_id and child span as span_id.

https://opentelemetry.io/docs/concepts/signals/traces/ 

```
{
  "name": "hello-greetings",
  "context": {
    "trace_id": "5b8aa5a2d2c872e8321cf37308d69df2",
    "span_id": "5fb397be34d26b51"
  },
  "parent_id": "051581bf3cb55c13",
  "start_time": "2022-04-29T18:52:58.114304Z",
  "end_time": "2022-04-29T22:52:58.114561Z",
  "attributes": {
    "http.route": "some_route2"
  }
```

**Reply by laurit** [reply]
There is no built-in support for adding parent span id to logs.
It might make things easier if you could change your ask a bit, perhaps instead the parent you want to log the id of the local root span or remote parent span?

**Reply by sulthans** [reply]
@laurit Thank you. How to log remote parent span in spring-boot? Currently i am working with Java Agent auto instrumentation
Shall i use extension or do we have other options, please suggest.

**Reply by trask** [reply]
hi @sulthans, just curious why you need the log's parent span's parent span?

**Reply by sulthans** [reply]
Hi @trask we have a application with multiple operations for a message received and the message may sent to another applications. When we search by traceId it will bring all operations/requests for the traceId now we are planning to log parent span id along with  traceId for the application specific operations and requests like apply one more filter with parent span id.
I am trying with extension example - DemoPropagatorProvider and DemoPropagator but the Autoservice is not loading, whether this help to add parent span id in context ?

DemoPropagator.java 

`@Override
  public <C> Context extract(Context context, C carrier, TextMapGetter<C> getter) {
    String traceParentValue = getter.get(carrier, TRACEPARENT_HEADER);
    System.out.printf("%d TraceParentValue from headers %n", traceParentValue);
    if (traceParentValue != null) {
      Matcher matcher = TRACEPARENT_PATTERN.matcher(traceParentValue);
      if (matcher.matches()) {
        String parentSpanId = matcher.group(1);
        System.out.printf("%d TraceParentValue after pattern applied %n", parentSpanId);
        return context.with(PARENT_SPAN_ID, parentSpanId);
      }
    }
    return context;
  }`

**Reply by trask** [reply]
you can already log the log's parent span id

it's only the log's parent span's parent span id that you can't log

**Comment by sulthans** [comment]
@trask I am not clear with log's parent span's parent span id, did you mean log's parent span id for local root span, log's parent span's parent span id for remote parent span

OTEL Exported json has remote parent span id.

Initial trace has parent span id - **051581bf3cb55c13**

`{
  "name": "hello",
  "context": {
    "trace_id": "5b8aa5a2d2c872e8321cf37308d69df2",
    "span_id": "051581bf3cb55c13"
  },
  "parent_id": null,
  "start_time": "2022-04-29T18:52:58.114201Z",
  "end_time": "2022-04-29T18:52:58.114687Z",
  "attributes": {
    "http.route": "some_route1"
  }
}`


Here for new operation new span and parent span id and the same set in parent_id - **051581bf3cb55c13** cannot log ?

`{
  "name": "hello-greetings",
  "context": {
    "trace_id": "5b8aa5a2d2c872e8321cf37308d69df2",
    "span_id": "5fb397be34d26b51"
  },
  "parent_id": "051581bf3cb55c13",
  "start_time": "2022-04-29T18:52:58.114304Z",
  "end_time": "2022-04-29T22:52:58.114561Z",
  "attributes": {
    "http.route": "some_route2"
  }
}`

Please confirm, Thank you.

**Reply by trask** [reply]
> OTEL Exported json has remote parent span id.

this is not true for logs, because logs aren't spans, so span_id on logs is really the log's parent span id

**Reply by sulthans** [reply]
Thanks @trask
